### PR TITLE
[ConstitutiveLawsApp] Reimplementing/Correcting Finite Strain Isotropic & Kinematic Plasticity

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm reimplementing carefully the isotropic and kinematic plasticity inside the CLA within the **Finite Strain framework**.

Test should be corrected in this PR before merging (incoming).

### CODE REVIEW

File: applications/ConstitutiveLawsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_kinematic_plasticity.h [hunk 1]
```diff
@@ -198,104 +199,86 @@ class GenericFiniteStrainConstitutiveLawIntegratorKinematicPlasticity
...
-            rConstitutiveLaw.CalculateValue(rValues, rStrainVariable, delta_plastic_strain);
-            // We compute the plastic strain
+            if (plastic_consistency_factor_increment < 0.0) plastic_consistency_factor_increment = 0.0;
+
+            // Decomposition Fe = Re*Ue
+            ConstitutiveLawUtilities<VoigtSize>::PolarDecomposition(rTrialElasticDeformationGradient, Re, Ue);
+
+            // Update Fe
+            noalias(rTrialElasticDeformationGradient) = ConstitutiveLawUtilities<VoigtSize>::
+                CalculateExponentialElasticDeformationGradient(rTrialElasticDeformationGradient,
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_kinematic_plasticity.h:240]: Why not one line? (80 character limit is sooo 80's)

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h [hunk 2]
```diff
@@ -366,6 +366,39 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ConstitutiveLawUtilities
...
+     * @param rFp The plastic def gradient tensor
+     */
+    static Matrix CalculateElasticDeformationGradient(
+        const MatrixType& rF,
+        const MatrixType& rFp
+        );
+    /**
+     * @brief This computes factorial of a number
+     */
+    static int Factorial(const unsigned int n) {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h]: This to the MathUtilities

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp [hunk 3]
```diff
@@ -870,6 +906,32 @@ Matrix ConstitutiveLawUtilities<TVoigtSize>::CalculateDirectPlasticDeformationGr
 /***********************************************************************************/
 /***********************************************************************************/
 
+template<SizeType TVoigtSize>
+void ConstitutiveLawUtilities<TVoigtSize>::CalculateExponentialOfMatrix(
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp]: This to the MathUtilities

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments and please, one line better than splitinglines

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h]: And an independent PR

**Comment by AlejandroCornejo** [comment]
> 
> 
> Some comments and please, one line better than splitinglines

haha The main thing is that with the one-line is impossible to read two files in parallel, for me is more efficient to control max width...

**Comment by loumalouomega** [comment]
> > Some comments and please, one line better than splitinglines
> 
> haha The main thing is that with the one-line is impossible to read two files in parallel, for me is more efficient to control max width...

In VS Code you can use Alt + Z to do dynamic lines (the thing is like that the code looks chaotic)

**Review by philbucher — CHANGES_REQUESTED** [review]
please move the functions to the CLawApp
Otherwise it will be even harder to eventually make the CLawApp independent of Structural

(same for the other functions in this file that are only used in the CLawsApp, but please different PR, and is not urgent)

**Comment by AlejandroCornejo** [comment]
> 
> 
> please move the functions to the CLawApp
> Otherwise it will be even harder to eventually make the CLawApp independent of Structural
> 
> (same for the other functions in this file that are only used in the CLawsApp, but please different PR, and is not urgent)

Basically would be moving the whole `constitutive_law_utilities.h`. The SMA needs this file?

**Comment by philbucher** [comment]
> > please move the functions to the CLawApp
> > Otherwise it will be even harder to eventually make the CLawApp independent of Structural
> > (same for the other functions in this file that are only used in the CLawsApp, but please different PR, and is not urgent)
> 
> Basically would be moving the whole `constitutive_law_utilities.h`. The `SMA `needs this file?

I think there are some things needed in the Stuctural.
My idea is to have a similar util in the CLawApp, which has the fcts needed by the advanced CLaws

If there is an overlap (functions needed by both apps) then I would move those to the core in the long run, then probably also others Apps could use them

**Comment by AlejandroCornejo** [comment]
> 
> 
> > > please move the functions to the CLawApp
> > > Otherwise it will be even harder to eventually make the CLawApp independent of Structural
> > > (same for the other functions in this file that are only used in the CLawsApp, but please different PR, and is not urgent)
> > 
> > 
> > Basically would be moving the whole `constitutive_law_utilities.h`. The `SMA `needs this file?
> 
> I think there are some things needed in the Stuctural.
> My idea is to have a similar util in the CLawApp, which has the fcts needed by the advanced CLaws
> 
> If there is an overlap (functions needed by both apps) then I would move those to the core in the long run, then probably also others Apps could use them

ok! I'll call it `advanced_constitutive_law_utilities.h`. I'll also move some methods to the MathUtils as @loumalouomega proposes

**Comment by AlejandroCornejo** [comment]
I'm adding the methods to `MathUtils` in the `adding-methods-to-Math-Utils` branch

**Comment by AlejandroCornejo** [comment]
@philbucher  I've created an `advanced_contitutive_law_utilities `class inside the CLApp and cleaned all the methods from the `constitutive_law_utils `that are only used inside the CLApp.

Factorial and exponential of a method are now inside the MathUtils.

The tests still must be updated

**Comment by AlejandroCornejo** [comment]
Now the checks are passing @loumalouomega , What do you guys think?

**Comment by loumalouomega** [comment]
I need to recheck...

**Comment by AlejandroCornejo** [comment]
> 
> 
> I need to recheck...

Sure, just wanted to let you know 👯‍♂️

**Review by philbucher — APPROVED** [review]
nice work!

I recommend checking codacy

File: applications/ConstitutiveLawsApplication/CMakeLists.txt [hunk 4]
```diff
@@ -16,6 +16,7 @@ file(GLOB_RECURSE KRATOS_CONSTITUTIVE_LAWS_APPLICATION_CORE
     ${CMAKE_CURRENT_SOURCE_DIR}/constitutive_laws_application.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/constitutive_laws_application_variables.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/custom_constitutive/*.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/custom_utilities/*.cpp
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/CMakeLists.txt:19]: Nice

File: applications/ConstitutiveLawsApplication/custom_constitutive/generic_small_strain_isotropic_plasticity.cpp [hunk 5]
```diff
@@ -105,7 +105,7 @@ void GenericSmallStrainIsotropicPlasticity<TConstLawIntegratorType>::CalculateMa
     } else { // We check for plasticity
         // Integrate Stress plasticity
         Vector& r_integrated_stress_vector = rValues.GetStressVector();
-        const double characteristic_length = ConstitutiveLawUtilities<VoigtSize>::CalculateCharacteristicLength(rValues.GetElementGeometry());
+        const double characteristic_length = AdvancedConstitutiveLawUtilities<VoigtSize>::CalculateCharacteristicLength(rValues.GetElementGeometry());
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/custom_constitutive/generic_small_strain_isotropic_plasticity.cpp:108]: Question, did we solved the issue of the characteristic length in the initial configuration?

**Review by loumalouomega — APPROVED** [review]
Noice

**Comment by loumalouomega** [comment]
Thanks @AlejandroCornejo

File: applications/ConstitutiveLawsApplication/custom_constitutive/generic_small_strain_isotropic_plasticity.cpp (hunk 5 above)
Comment (loumalouomega) [reply applications/ConstitutiveLawsApplication/custom_constitutive/generic_small_strain_isotropic_plasticity.cpp:108]: @AlejandroCornejo ping, did we solved the issue?

Comment (AlejandroCornejo) [reply applications/ConstitutiveLawsApplication/custom_constitutive/generic_small_strain_isotropic_plasticity.cpp:108]: Now we have the two methods... You can compute the char lenght in the original or updated conf

**Comment by ddiezrod** [comment]
@jrubiogonzalez I think we will have to do some changes in our side when you do the next merge after this MR

**Comment by AlejandroCornejo** [comment]
> 
> 
> @jrubiogonzalez I think we will have to do some changes in our side when you do the next merge after this MR

Do you need any help? :D

**Comment by ddiezrod** [comment]
> > @jrubiogonzalez I think we will have to do some changes in our side when you do the next merge after this MR
> 
> Do you need any help? :D

I  hope not :S

**Comment by loumalouomega** [comment]
I don't want to be carrier of bad news, but nightly CL test fail: https://github.com/KratosMultiphysics/Kratos/runs/2968375660?check_suite_focus=true

**Comment by AlejandroCornejo** [comment]
> 
> 
> I don't want to be carrier of bad news, but nightly CL test fail: https://github.com/KratosMultiphysics/Kratos/runs/2968375660?check_suite_focus=true

I'll look into it

**Pull request merged**